core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java

```diff
@@ -358,10 +358,20 @@ public long spill(long numBytes) throws IOException {
           return 0L;
         }
 
+        long released = 0L;
+        // longArray will not be used anymore, but may take significant amount of memory
+        // but never released until the task is complete.
+        if (longArray != null) {
+          released += longArray.memoryBlock().size();
+          freeArray(longArray);
+          longArray = null;
+        }
+        if (released >= numBytes) {
+          return released;
```

Comment:
AFAIK, `spill` is called when other consumers want to allocate in-memory pages but can't get any memory. If we are able to release ample memory by free'ing up the `longArray`, then the purpose of `spill` is achieved. Hence the short circuit.